src/Store/MongoDB/Store.php

```diff
@@ -0,0 +1,133 @@
+     * @param string $collectionName  The name of the collection
+     * @param string $indexName       The name of the index
+     * @param string $vectorFieldName The name of the field int the index that contains the vector
+     */
+    public function __construct(
+        private Client $client,
+        private LoggerInterface $logger,
+        private string $databaseName,
+        private string $collectionName,
+        private string $indexName,
+        private string $vectorFieldName = 'vector',
+    ) {
+    }
+
+    public function addDocument(Document $document): void
+    {
+        $this->addDocuments([$document]);
+    }
+
+    public function addDocuments(array $documents): void
+    {
+        $collection = $this->client->selectCollection($this->databaseName, $this->collectionName);
+
+        foreach ($documents as $document) {
+            if (!$document->hasVector()) {
+                $this->logger->warning('Document {id} does not have a vector', ['id' => $document->id]);
+            }
+
+            $id = $document->id->toRfc4122();
+
+            $collection->replaceOne(
+                ['id' => $id],
+                array_filter([
+                    'metadata' => $document->metadata->getArrayCopy(),
+                    $this->vectorFieldName => $document->vector->getData(),
+                    'id' => $id,
+                    'text' => $document->text,
+                ]),
+                ['upsert' => true], // insert if not exists
+            );
+        }
+    }
+
+    /**
+     * @param array{
+     *     limit?: positive-int,
+     *     numCandidates?: positive-int,
+     * } $options
+     *
+     * @return array|Document[]
+     */
+    public function query(Vector $vector, array $options = []): array
+    {
+        $collection = $this->client->selectCollection($this->databaseName, $this->collectionName);
+
+        $results = $collection->aggregate([
+            [
+                '$vectorSearch' => [
+                    'index' => $this->indexName,
+                    'path' => $this->vectorFieldName,
+                    'queryVector' => $vector->getData(),
+                    'numCandidates' => $options['numCandidates'] ?? 200,
+                    'limit' => $options['limit'] ?? 5,
+                ],
+            ],
+            [
+                '$project' => [
+                    '_id' => 0,
+                    'id' => 1,
+                    'text' => 1,
+                    'metadata' => 1,
+                    $this->vectorFieldName => 1,
+                    'score' => [
+                        '$meta' => 'vectorSearchScore',
+                    ],
+                ],
+            ],
+        ]);
+
+        $documents = [];
+        foreach ($results as $result) {
+            $result = $result->getArrayCopy();
+
+            $documents[] = Document::fromVector(
+                Vector::create1536($result[$this->vectorFieldName]),
+                Uuid::fromString($result['id']),
```

Comment:
Noticed that you're storing a UUID as the identifier. Consider storing this as a [binary field](https://www.php.net/manual/en/class.mongodb-bson-binary.php#mongodb-bson-binary.constants.type-uuid) for better storage efficiency

Reply:
After you made these changes, I looked for my previous work with the BSON binary type and Uuid:
```php
use MongoDB\BSON\Binary;
use Symfony\Component\Uid\Uuid;

function toBinary(Uuid $uuid): Binary
{
    return new Binary($uuid->toBinary(), Binary::TYPE_UUID);
}

function toUuid(Binary $binary): Binary
{
    return Uuid::fromString($binary->getData());
}
```

Reply:
done